File: dotnet/fsharp/Pxl.Ui/Evaluation.fs
namespace Pxl

open System
open System.Threading
open Pxl

type Reality
    (
        onCycleFinished: DateTimeOffset -> unit,
        [<InlineIfLambda>] getNow: unit -> DateTimeOffset
    )
    =
    member _.OnCycleFinished(nextPlannedEvaluation: DateTimeOffset) =
        onCycleFinished nextPlannedEvaluation
    member _.Now =
        getNow()

module Reality =
    let forRealTime () =
        let getNow () = DateTimeOffset.Now
        let onCycleFinished nextPlannedEvaluation =
                let timeToWait = nextPlannedEvaluation - getNow ()
                if timeToWait > TimeSpan.Zero then
                    Thread.Sleep(timeToWait)
        Reality(onCycleFinished, getNow)

// TODO: ggf. eine HandleErrorStrategy aus onError machen
module Evaluation =
    let mutable startedTimes = 0

    let hangDetectionTimeSpan = TimeSpan.FromSeconds(5.0)

    let start
        (
            canvas: Canvas,
            onEvalError: Exception -> unit,
            reality: Reality,
            readButtons: unit -> Buttons,
            scene: Vide<unit,'s>
        ) =
        let mutable shouldEvaluate = true
        let isRunning () = shouldEvaluate && not canvas.Ct.IsCancellationRequested

        let mutable lastEvaluationTime = None

        fun () ->
            let frameArrays =
                [
                    for i in 0 .. canvas.SendBufferSize - 1 do
                        Array.zeroCreate<Color>(canvas.Metadata.width * canvas.Metadata.height)
                ]
            let renderCtx = RenderCtx(canvas.Metadata.width, canvas.Metadata.height, canvas.Metadata.fps)
            let durationForOneFrame = 1.0 / float canvas.Metadata.fps
            let sceneStartTime = reality.Now
            let mutable lastSceneState = None
            let mutable completeCycleCount = 0
            let calcTimeForCycle cycleNr = sceneStartTime.AddSeconds(durationForOneFrame * float cycleNr)
            while isRunning () do
                do lastEvaluationTime <- Some reality.Now
                try
                    // Für die Szene ist es wichtig, dass "now" keinen Jitter hat.
                    // Wir berechnen jeden Zyklus - egal, wie weit wir hintendran sind.
                    // Die Puffer gleichen das wieder aus durh Frame-Dropping im härtesten Fall.
                    // Im Schnitt ist der RPi schon stark genug, um wieder aufzuholen.
                    let frameNow = calcTimeForCycle completeCycleCount
                    let frame =
                        do renderCtx.PrepareCycle(sceneStartTime, frameNow, readButtons ())
                        do lastSceneState <- scene lastSceneState renderCtx |> snd
                        let frame = frameArrays[completeCycleCount % frameArrays.Length]
                        do renderCtx.EndCycle(frame)
                        frame
                    do
                        canvas.PushFrameSafe(frame)
                        completeCycleCount <- completeCycleCount + 1
                        reality.OnCycleFinished(calcTimeForCycle completeCycleCount)
                with ex ->
                    printfn $"Error in evaluation: {ex.Message}"
                    onEvalError ex
        |> Thread.startBackground $"Evaluation_{startedTimes}"

        fun () ->
            while isRunning () do
                match lastEvaluationTime with
                | Some lastEvaluationTime ->
                    if reality.Now - lastEvaluationTime > hangDetectionTimeSpan then
                        onEvalError <| new Exception("Evaluation hanging")
                | None -> ()
                Thread.Sleep(100)
        |> Thread.startBackground $"EvaluationHangDetection_{startedTimes}"

        fun () -> shouldEvaluate <- false

File: node/typescript/pxl-ui/src/evaluation.ts
import { Canvas } from "./canvas.js";
import { Color } from "./color.js";
import { RenderCtx } from "./renderCtx.js";
import { ResumableStateHack, SceneFunc } from "./vide.js";

// Reality class and module
export class Reality {
  constructor(
    public readonly onCycleFinished: (nextPlannedEvaluation: Date) => Promise<void>,
    private readonly getNow: () => Date
  ) { }

  get now(): Date {
    return this.getNow();
  }

  static forRealTime(): Reality {
    const getNow = () => new Date();

    // Instead of Thread.Sleep, we return a promise that resolves after the given delay.
    const onCycleFinished = async (nextPlannedEvaluation: Date) => {
      const now = getNow().getTime();
      const target = nextPlannedEvaluation.getTime();
      const timeToWait = target - now;
      if (timeToWait > 0) {
        await new Promise<void>((resolve) => setTimeout(resolve, timeToWait));
      }
    };

    return new Reality(onCycleFinished, getNow);
  }
}

// Evaluation module
let startedTimes = 0;
const hangDetectionTimeSpan = 5000; // 5 seconds

let stop: () => void = () => { }

// The 'start' function has been adapted to use setTimeout and async loops instead of threads.
export function startEx<S>(
  canvas: Canvas,
  onEvalError: (ex: Error) => void,
  reality: Reality,
  // readButtons: () => Buttons,
  scene: SceneFunc<S>
)
  : { stop: () => void } {

  stop();

  startedTimes++;
  let shouldEvaluate = true;
  const isRunning = () => shouldEvaluate && !canvas.isCancellationRequested;

  let lastEvaluationTime: Date | null = null;

  // Prepare frames
  const frameArrays: Color[][] = [];
  for (let i = 0; i < canvas.sendFrameBufferSize; i++) {
    frameArrays.push(new Array<Color>(canvas.metadata.width * canvas.metadata.height));
  }

  const renderCtx = new RenderCtx(canvas.metadata.width, canvas.metadata.height);
  const durationForOneFrame = 1.0 / canvas.metadata.fps;
  const sceneStartTime = reality.now;

  let lastSceneState: readonly any[] | undefined = undefined;
  let completeCycleCount = 0;

  const calcTimeForCycle = (cycleNr: number) =>
    new Date(sceneStartTime.getTime() + (durationForOneFrame * cycleNr * 1000));

  // The main evaluation loop as an async function that schedules itself
  async function evaluationLoop() {
    if (!isRunning()) return;

    try {
      lastEvaluationTime = reality.now;

      // Compute frame time based on ideal schedule
      const frameNow = calcTimeForCycle(completeCycleCount);

      renderCtx.prepareCycle(frameNow);
      ResumableStateHack.popStack = lastSceneState;
      ResumableStateHack.pushStack = [];
      ResumableStateHack.ctx = renderCtx;

      await scene();

      lastSceneState = ResumableStateHack.pushStack;
      ResumableStateHack.popStack = undefined;
      ResumableStateHack.pushStack = [];
      ResumableStateHack.ctx = undefined;

      const frame = renderCtx.endCycle();

      canvas.pushFrame(frame);
      completeCycleCount++;

      const nextPlannedEvaluation = calcTimeForCycle(completeCycleCount);
      await reality.onCycleFinished(nextPlannedEvaluation);

    } catch (ex: any) {
      console.log(`Error in evaluation: ${ex.message}`);
      onEvalError(ex);
    }

    if (isRunning()) {
      // Schedule next frame evaluation using setTimeout with minimal delay
      // Actual timing is managed by reality.OnCycleFinished
      setTimeout(evaluationLoop, 0);
    }
  }

  // Hang detection loop: checks periodically if the evaluation is hanging
  const hangDetectionInterval = setInterval(() => {
    if (!isRunning()) return;
    if (lastEvaluationTime) {
      const now = reality.now.getTime();
      const lastEval = lastEvaluationTime.getTime();
      if (now - lastEval > hangDetectionTimeSpan) {
        onEvalError(new Error("Evaluation hanging"));
      }
    }
  }, 100);

  // Start the evaluation loop
  setTimeout(evaluationLoop, 0);

  // Return a function that stops the evaluation
  stop = () => {
    shouldEvaluate = false;
    clearInterval(hangDetectionInterval);
  };

  return { stop };
}

export function start(canvas: Canvas, scene: SceneFunc<void>) {
  const reality = Reality.forRealTime();
  return startEx(
    canvas,
    err => console.log("Error in evaluation", err.message),
    reality,
    scene);
}
